controllers/drain_controller.go

```diff
@@ -0,0 +1,464 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	sriovnetworkv1 "github.com/k8snetworkplumbingwg/sriov-network-operator/api/v1"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/drain"
+	constants "github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/global/consts"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/global/vars"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/platforms"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/utils"
+)
+
+type DrainReconcile struct {
+	client.Client
+	Scheme         *runtime.Scheme
+	drainer        drain.DrainInterface
+	resourcePrefix string
+
+	nodesInReconcile      map[string]interface{}
+	nodesInReconcileMutex sync.Mutex
+	drainCheckMutex       sync.Mutex
+}
+
+func NewDrainReconcileController(client client.Client, Scheme *runtime.Scheme, platformHelper platforms.PlatformHelperInterface) (*DrainReconcile, error) {
+	resourcePrefix := os.Getenv("RESOURCE_PREFIX")
+	if resourcePrefix == "" {
+		return nil, fmt.Errorf("RESOURCE_PREFIX environment variable can't be empty")
+	}
+
+	drainer, err := drain.NewDrainer(resourcePrefix, platformHelper)
+	if err != nil {
+		return nil, err
+	}
+
+	return &DrainReconcile{
+		client,
+		Scheme,
+		drainer,
+		resourcePrefix,
+		map[string]interface{}{},
+		sync.Mutex{},
+		sync.Mutex{}}, nil
+}
+func (dr *DrainReconcile) TryLockNode(nodeName string) bool {
+	dr.nodesInReconcileMutex.Lock()
+	defer dr.nodesInReconcileMutex.Unlock()
+
+	_, exist := dr.nodesInReconcile[nodeName]
+	if exist {
+		return false
+	}
+
+	dr.nodesInReconcile[nodeName] = nil
+	return true
+}
+
+func (dr *DrainReconcile) unlockNode(nodeName string) {
+	dr.nodesInReconcileMutex.Lock()
+	defer dr.nodesInReconcileMutex.Unlock()
+
+	delete(dr.nodesInReconcile, nodeName)
+}
+
+//+kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=sriovnetwork.openshift.io,resources=sriovnodestates,verbs=get;list;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (dr *DrainReconcile) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// try to lock the node to this reconcile loop.
+	// if we are not able this means there is another loop already handling this node, so we just exist
```

Comment:
When would that happen? When would 2 reconciles be running for the same node?

Reply:
right, and also by the operator itself as we apply changes to the annotation we will get a new reconcile call so we just drop them

Reply:
I think  that implementation of the queue in the client-go package [1] will protect you from this case. You will have only 1 reconcile request for one key(node name in our case) even if max parallel reconciles is 50.

[1] https://github.com/adrianchiris/sriov-network-operator/blob/56530b9c0e1c893d686f545df6876ddce807950f/vendor/k8s.io/client-go/util/workqueue/queue.go#L163